UNL-CSE-AMBASSADORS/VITA
# Limit volunteers on shifts

## Overview
This PR enables the functionality of limiting the number of volunteers who can sign up for a certain role at a site/shift. 

## Testing
1. Run the database scripts
2. Log in as preparer
3. Navigate to profile page
4. Scroll down to the shift sign up section
5. Hit the "Add a shift" button
    * Select "Test Site" -> 1st date option -> only time option
        * When you go to select a role, it should say there is a 10 person limit on the preparer role
    * Sign up for that shift in the preparer role
6. Hit the "Add a shift" button
    * Select "Test Site" -> 2nd date option -> only time option
        * When you go to select a role, it should say there is a 15 person limit on the preparer role
    * Sign up for that shift in the preparer role
7. Hit the "Add a shift" button
    * All the selects except the site select should be disabled
    * Once you select a site, the date select should be enabled, and the other two should be disabled
    * Once you select a date, the time select should be enabled, and the role select should be disabled
    * If you change the site/date/time the selects to the right should reset (and become disabled if applicable)
8. Log out and sign in as siteadmin
9. Navigate to the profile page
10. Hit the "Add a shift" button
    * Select "Test Site" -> 1st date option -> only time option
        * It should say there is a 10 person limit on the preparer role, and it should say there are 9 spots remaining
    * Select the 2nd date option -> only time option
        * It should say there is a 15 person limit on the preparer role, and it should say there are 13 spots remaining

## Production database evolution
The following script will need to be run against the production database when this feature is deployed.
```
CREATE TABLE RoleLimit (
    roleLimitId INTEGER UNSIGNED PRIMARY KEY NOT NULL AUTO_INCREMENT,
    maximumNumber INTEGER UNSIGNED NOT NULL,
    roleId INTEGER UNSIGNED NOT NULL,
    FOREIGN KEY(roleId) REFERENCES Role(roleId),
    siteId INTEGER UNSIGNED NOT NULL,
    FOREIGN KEY(siteId) REFERENCES Site(siteId),
    shiftId INTEGER UNSIGNED NULL,
    FOREIGN KEY(shiftId) REFERENCES Shift(shiftId),
    CONSTRAINT uniqueRoleSiteShiftTuple UNIQUE INDEX(roleId, siteId, shiftId)
);
```

## New Database ERD
![databaseerd](https://user-images.githubusercontent.com/9295744/48327784-41a00b00-e606-11e8-87d3-a4830d0f0ee7.png)
